<!-- data flow  -->
<!-- audio -> spectrum -> filtered spectrum -> spectrogram -->

<launch>

  <arg name="launch_audio_capture" default="true" />

  <arg name="bitdepth" default="16" />
  <arg name="mic_sampling_rate" default="16000" />
  <arg name="n_channel" default="1" />
  <arg name="device" default="hw:0,0" />
  <arg name="audio_topic" default="/audio" />
  <arg name="fft_sampling_period" default="0.3" />
  <arg name="fft_exec_rate" default="50" />
  <arg name="high_cut_freq" default="800" />
  <arg name="low_cut_freq" default="1" />  <!-- default: remove 0Hz because this component is meaningless in audio -->
  <arg name="spectrogram_period" default="5" />
  <arg name="gui" default="true" />

  <!-- Publish audio topic from microphone -->
  <node name="audio_capture" pkg="audio_capture" type="audio_capture"
        if="$(arg launch_audio_capture)"
        respawn="true">
    <rosparam subst_value="true">
      format: wave
      channels: $(arg n_channel)
      depth: $(arg bitdepth)
      sample_rate: $(arg mic_sampling_rate)
      device: $(arg device)
    </rosparam>
  </node>

  <!-- convert audio topic to spectrum topic -->
  <node name="audio_to_spectrum"
        pkg="audio_to_spectrogram" type="audio_to_spectrum.py"
        respawn="true">
    <remap from="~audio" to="$(arg audio_topic)" />
    <rosparam subst_value="true">
      n_channel: $(arg n_channel)
      mic_sampling_rate: $(arg mic_sampling_rate)
      fft_sampling_period: $(arg fft_sampling_period)
      bitdepth: $(arg bitdepth)
      fft_exec_rate: $(arg fft_exec_rate)
    </rosparam>
  </node>

  <!-- filter log spectrum -->
  <node name="log_spectrum_filter"
        pkg="audio_to_spectrogram" type="spectrum_filter.py"
        respawn="true">
    <remap from="~input" to="audio_to_spectrum/log_spectrum" />
    <rosparam subst_value="true">
      data_sampling_rate: $(arg mic_sampling_rate)
      high_cut_freq: $(arg high_cut_freq)
      low_cut_freq: $(arg low_cut_freq)
    </rosparam>
  </node>

  <!-- publish amplitude image -->
  <node name="audio_amplitude_plot"
        pkg="audio_to_spectrogram" type="audio_amplitude_plot.py"
        output="screen">
    <remap from="~audio" to="$(arg audio_topic)" />
    <rosparam subst_value="true">
      n_channel: $(arg n_channel)
      mic_sampling_rate: $(arg mic_sampling_rate)
      bitdepth: $(arg bitdepth)
    </rosparam>
  </node>

  <group if="$(arg gui)">
    <node pkg="image_view" type="image_view" name="audio_amplitude_view"  >
      <remap from="image" to="audio_amplitude_plot/output/viz" />
    </node>
  </group>

  <include file="$(find audio_to_spectrogram)/launch/spectrum_plot.launch">
    <arg name="spectrum" value="log_spectrum_filter/output" />
    <arg name="gui" value="$(arg gui)" />
  </include>

  <include file="$(find audio_to_spectrogram)/launch/spectrum_to_spectrogram.launch">
    <arg name="spectrum" value="log_spectrum_filter/output" />
    <arg name="spectrogram_period" value="$(arg spectrogram_period)" />
    <arg name="gui" value="$(arg gui)" />
  </include>

</launch>
